### PR TITLE
Restructuring SoundFont Structs

### PR DESCRIPTION
**Pull request by engineer124**
A follow-up to the previous soundFonts PR https://github.com/zeldaret/oot/pull/1237. This organizes the structs used in soundFonts together, although I stopped one step short of making a separate `soundFonts.h` header (open to opinions on this).

Some of the following changes includes:
- `SoundFontSample` -> `SampleInfo`, the main struct that contains meta-data to an individual struct
- `SoundFontSound` -> `TunedSampleInfo`, a struct that contains `SampleInfo` but with an extra tuning parameter
- Introducing a separate `SoundEffects` struct to parallel `Instruments` and `Drums` so that meta-data for `SoundEffects` can be modified without depending on the underlying `TunedSampleInfo` struct
- `SoundFont` -> `SoundFontInfo`, a struct not contained within soundFonts, but a struct found on the audio heap (in the init pool, see https://github.com/zeldaret/oot/pull/1273) used to store parsed information from a soundFont so that the full soundFont data only needs to be read from once
- `highNotesSound` -> `highPitchTunedSample` and same for normal/low
- Make the union for the pointer to the tunedSample/waveSample in NoteSubEu anonymous
- Update function names to match the new structs

### CODE REVIEW

File: src/code/audio_seqplayer.c [hunk 1]
```diff
@@ -769,27 +769,28 @@ s32 AudioSeq_SeqLayerProcessScriptStep4(SequenceLayer* layer, s32 cmd) {
...
+            layer->tunedSample = tunedSample;
+            layer->freqScale = tunedSample->tuning;
             break;
 
         case 1:
             layer->semitone = semitone;
             sfxId = (layer->transposition << 6) + semitone;
-            sound = Audio_GetSfx(channel->fontId, sfxId);
-            if (sound == NULL) {
+            soundEffects = Audio_GetSfx(channel->fontId, sfxId);
```
Comment (Dragorn421) [inline src/code/audio_seqplayer.c:785]:
```suggestion
            soundEffect = Audio_GetSfx(channel->fontId, sfxId);
```

File: src/code/audio_heap.c [hunk 2]
```diff
@@ -1440,30 +1440,30 @@ void AudioHeap_ApplySampleBankCacheInternal(s32 apply, s32 sampleBankId) {
...
+                    AudioHeap_ChangeStorage(&change, drum->tunedSample.sample);
                 }
             }
 
-            for (sfxId = 0; sfxId < gAudioContext.soundFonts[fontId].numSfx; sfxId++) {
-                sfx = Audio_GetSfx(fontId, sfxId);
-                if (sfx != NULL) {
-                    AudioHeap_ChangeStorage(&change, sfx->sample);
+            for (sfxId = 0; sfxId < gAudioContext.soundFontInfoList[fontId].numSfx; sfxId++) {
+                soundEffects = Audio_GetSfx(fontId, sfxId);
```
Comment (Dragorn421) [inline src/code/audio_heap.c:1464]:
```suggestion
                soundEffect = Audio_GetSfx(fontId, sfxId);
```

File: src/code/audio_load.c [hunk 3]
```diff
@@ -1297,30 +1298,30 @@ s32 AudioLoad_SlowLoadSample(s32 fontId, s32 instId, s8* status) {
...
         Drum* drum = Audio_GetDrum(fontId, instId - 0x80);
 
         if (drum == NULL) {
             return NULL;
         }
-        sample = drum->sound.sample;
+        sample = drum->tunedSample.sample;
     } else {
-        SoundFontSound* sound = Audio_GetSfx(fontId, instId - 0x100);
+        SoundEffect* soundEffects = Audio_GetSfx(fontId, instId - 0x100);
```
Comment (Dragorn421) [inline src/code/audio_load.c:1319]:
```suggestion
        SoundEffect* soundEffect = Audio_GetSfx(fontId, instId - 0x100);
```

File: src/code/audio_load.c [hunk 4]
```diff
@@ -916,10 +917,10 @@ void AudioLoad_RelocateFont(s32 fontId, SoundFontData* fontDataStartAddr, Sample
 
 #undef FONT_DATA_RELOC
 
-    // Store the relocated pointers to the sound lists in the soundFonts meta-data struct
-    gAudioContext.soundFonts[fontId].drums = (Drum**)fontData[0];
-    gAudioContext.soundFonts[fontId].soundEffects = (SoundFontSound*)fontData[1];
-    gAudioContext.soundFonts[fontId].instruments = (Instrument**)(fontData + 2);
+    // Store the relocated pointers to the soundFontInfoList list
```
Comment (Dragorn421) [inline src/code/audio_load.c:920]:
```suggestion
    // Store the relocated pointers
```
:thinking:

File: src/code/audio_load.c [hunk 5]
```diff
@@ -800,15 +801,15 @@ AudioTable* AudioLoad_GetLoadTable(s32 tableType) {
  * @param sampleBankReloc information on the sampleBank containing raw audio samples
  */
 void AudioLoad_RelocateFont(s32 fontId, SoundFontData* fontDataStartAddr, SampleBankRelocInfo* sampleBankReloc) {
-    u32 soundOffset;     // Relative offset from the beginning of fontData directly to the sound/envelope
+    u32 soundOffset;     // Relative offset from the beginning of fontData directly to the tunedSample/envelope
     u32 soundListOffset; // Relative offset from the beginning of fontData to the list of soundOffsets/sfxs
```
Comment (Dragorn421) [inline src/code/audio_load.c:805]: not sure why I didn't catch this earlier but `soundOffset` can also be ram `soundOffset = RELOC_TO_RAM(soundOffset);`

and both of these are also used for drums

NotLikeThis

File: src/code/audio_load.c [hunk 6]
```diff
@@ -800,15 +801,15 @@ AudioTable* AudioLoad_GetLoadTable(s32 tableType) {
...
     Instrument* inst;
     Drum* drum;
-    SoundFontSound* sfx;
+    TunedSampleInfo* sfx;
```
Comment (Dragorn421) [inline src/code/audio_load.c:808]:
```suggestion
    SoundEffect* soundEffect;
```
right?

File: src/code/audio_load.c [hunk 7]
```diff
@@ -862,14 +863,14 @@ void AudioLoad_RelocateFont(s32 fontId, SoundFontData* fontDataStartAddr, Sample
         // Loop through the sound effects
         for (i = 0; i < numSfx; i++) {
             // Get a pointer to the i'th sound effect
-            soundOffset = ((SoundFontSound*)fontData[1]) + i;
+            soundOffset = ((TunedSampleInfo*)fontData[1]) + i;
 
             // Check for NULL (note: the pointer is guaranteed to be in fontData and can never be NULL)
-            if ((SoundFontSound*)soundOffset != NULL) {
+            if ((TunedSampleInfo*)soundOffset != NULL) {
```
Comment (Dragorn421) [inline src/code/audio_load.c:869]: should be `SoundEffect` ?

File: src/code/code_800E4FE0.c [hunk 8]
```diff
@@ -517,9 +517,9 @@ u8* func_800E5E84(s32 arg0, u32* arg1) {
     return AudioLoad_GetFontsForSequence(arg0, arg1);
 }
 
-void func_800E5EA4(s32 arg0, u32* arg1, u32* arg2) {
-    *arg1 = gAudioContext.soundFonts[arg0].sampleBankId1;
-    *arg2 = gAudioContext.soundFonts[arg0].sampleBankId2;
+void Audio_GetSampleBankIds(s32 fontId, u32* sampleBankId1, u32* sampleBankId2) {
```
[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/code_800E4FE0.c:520]: Unused yeah. Might as well go the whole mile and just say:
`Audio_GetSampleBankIdsOfSoundFont`

[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/code_800E4FE0.c:520]: Actually, `Audio_GetSampleBankIdsOfFont` is fine. Other functions use `Font` in its name so its consistent

[the remark this answers is not in the document]
Comment (engineer124) [reply src/code/code_800E4FE0.c]: Minor detail, there's a function `AudioLoad_GetSamplesForFont`, so for consistency: `ForFont` or `OfFont`?

[the remark this answers is not in the document]
Comment (Dragorn421) [reply src/code/code_800E4FE0.c]: there are also functions named `GetInstrumentSomething` instead of `getSomethingOfInstrument` :shrug: , as you want

**Review by Dragorn421 — APPROVED** [review]

**Review by fig02 — APPROVED** [review]

**Review by Roman971 — APPROVED** [review]

**Pull request merged**